[SILOpt] Handle reborrows of owned phis in CanonicalizeOSSALifetime.

nate-chandler opened the pull request:
When canonicalizing lifetime of an owned phi, update liveness for all reborrows any of whose underlying values are transitively borrows of any of the owned phi's underlying values.

rdar://94346482

nate-chandler committed:
[SIL] Add utility to transitively visit operands.

The new utility looks through operands whose values are themselves
SILPhiArguments and visits those arguments' operands.

nate-chandler committed:
[OwnershipUtils] Visit owned phi's reborrows.

The new utility, given an phi, visits all adjacent phis (i.e. arguments
to the same block) which are (potentially iterated) reborrows of a value
reaching the given phi.

nate-chandler committed:
[CanonicalizeOSSALifetime] Handle adj reborrows.

When canonicalizing lifetime of an owned phi, add adjacent phis which
are reborrows of its reaching values to the worklist.  Update liveness
to consider uses whose operand ownership is specific to guaranteed
values (i.e. uses which are derived from these adjacent reborrow phis)
as standard, non-lifetime-ending uses.

rdar://94346482

nate-chandler commented:
@swift-ci please test

atrick reviewed (changes requested):
Fantastic code and comments.

My concerns so far (I haven't had time to think of alternatives):

We're repeatedly recomputing phi webs and rediscovering the mapping between all the enclosing scope phis and inner scope phis

The PrunedLiveness abstraction is broken if we allow reborrows. We're either computing SSA liveness or phi-web liveness and that needs to be clear when we handle reborrows. For phi webs, we need to explicitly initialize liveness for all the roots of the phi-web as defs. If we're doing liveness for phi-webs, then we need to make it clear how to use the API correctly.

Along the same lines, we might be visiting uses that aren't dominated by the original def for which we initialized liveness.

We might want a general enclosingScope analysis that maps inner phis to their enclosing scope phis, but I'm not sure we need it yet.

nate-chandler commented on lib/SIL/Utils/OwnershipUtils.cpp:
```diff
@@ -1586,6 +1586,138 @@ void swift::findTransitiveReborrowBaseValuePairs(
+/// operands adjacent to the incoming operands to the owned phi argument and
+/// which are borrows of the value consumed there.  The reason is that they
+/// might not be borrows of that incoming owned value _directly_ but rather
+/// reborrows of some underlying def if the incoming owned value is itself a
```
Reworded in terms of {reaching values of, values reaching} the phi of interest.

nate-chandler commented on lib/SIL/Utils/OwnershipUtils.cpp:
```diff
@@ -1586,6 +1586,138 @@ void swift::findTransitiveReborrowBaseValuePairs(
+
+  // First, collect all the arguments that forward into arg and all the
+  // underlying values which it takes on along different control flow paths.
+  SmallVector<SILPhiArgument *, 4> ownedArguments;
```
Renamed this and other variables, hopefully to be clearer.

nate-chandler commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -1217,6 +1217,14 @@ void findTransitiveReborrowBaseValuePairs(
+///
+/// If the visitor returns false, stops visiting and returns false.  Otherwise,
+/// returns true.
+bool visitReborrowsOfOwnedPhiArgument(SILPhiArgument *arg,
```
Great point!  Renamed and dropped the assertion.

nate-chandler commented on lib/SIL/Utils/PrunedLiveness.cpp:
```diff
@@ -106,7 +106,8 @@ void PrunedLiveness::updateForUse(SILInstruction *user, bool lifetimeEnding) {
 bool PrunedLiveness::updateForBorrowingOperand(Operand *op) {
-  assert(op->getOperandOwnership() == OperandOwnership::Borrow);
+  assert(op->getOperandOwnership() == OperandOwnership::Borrow ||
+         op->getOperandOwnership() == OperandOwnership::Reborrow);
```
Removed this bit, now that we're adding the adjacent reborrow phis to the def use worklist.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
Awesome!

nate-chandler commented:
@swift-ci please test macOS platform

nate-chandler committed:
[CanonicalizeOSSALifetime] Handle later reborrows.

Properly handle the different sorts of reborrows when they are visited:
- non-branch reborrows (tuples, etc) are non-lifetime-ending uses; such
  reborrows are relevant defs for liveness
- branch reborrows which also consume the owned phi are lifetime-ending
  uses
- branch reborrows which DO NOT consume the owned phi are
  non-lifetime-ending uses, and the corresponding phis are relevant defs
  for liveness

rdar://97399065

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test macOS platform

nate-chandler commented:
@swift-ci please test linux platform
